### PR TITLE
Mega Refacto

### PR DESCRIPTION
**Pull request by JulienTant**
#### Summary
Sorry for this huge PR :)

**What is missing/broken:**
- Access control.
- Updating on the `Makefile` etc to compile the plugins. For the bash plugin for example the script is `go build -buildmode=plugin -o plugins/bash.so ./plugin/bash/*.go` but I don't know how to make  all the makefile magic 😅 cc @phoinixgrr 👀 
- I did not go through every single bash script to make sure they work, there's probably some work necessary on commands that have arguments

**What has changed:**
- Almost everything 😅 Sorry! :D 
- I tried to do a bit more decoupling overall, try to have things in place where they make sense? I hope it's gonna help us make the tool easy to maintain and improve. For example, I started working on the idea of the store for the dialogs session, right now it's in memory but in the future we could use something that is  saved between restarts etc.
- `plugins` provide commands, that we can associate with a "slash command". When configuring a command, we can filter what command we want to load from the plugins:
    - `only: []string`: only load the commands listed here
    - `exclude: []string`: exclude commands listed here
    - note: you can't use only+exclude at the same time
 
  This will allow us to control more precisely what is loaded from where and apply ACL on certain sets more easily. plugins can be used in multiple slash commands.
- Command Arguments: all commands must be double dashed flags (`--myarg=value` | `--myarg="my value"` | `--myarg` (value will default to the string `"true"` when no value is passed)). commands will receive an `map[string]string` will all parsed arguments. For bash specifically, we create a new env variable prefixed by `ARG_` for each argument, so `--myarg="hello"` is equivalent to `export ARG_MYARG="hello"` (see code bellow)
  ```go
  for name, value := range args {
    cmd.Env = append(cmd.Env, fmt.Sprintf("ARG_%s=%s", strings.ToUpper(name), value))
  }
  ```
- Logging: I swapped golang logger for a more robust and configurable solution by using Uber's zap (https://github.com/uber-go/zap) with context awareness for log metadata
- And probably much more 😅 

Feel free to ping me for any question or remark on my refacto, happy to work on it to make it better with your suggestions!

### CODE REVIEW

File: config/config.sample.yaml [hunk 1]
```diff
@@ -1,19 +1,47 @@
...
 listen: "0.0.0.0:8080"
-token: "TOKEN"           # Mattermost slash command token
-commands:
-  - commands/gitlab/gitlab.yaml                
-  - commands/docker/docker.yaml
-  - commands/fun/fun.yaml
-  - commands/k8s/k8s.yaml
-  - commands/release/release.yaml
+base_url: "https://op-tools-address.com"
+plugins:
```
Comment (pfltdv) [inline config/config.sample.yaml:5]: Great work @JulienTant. I love the seperation of plugins and command filters.

File: model/command.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import "context"
+
+type Command struct {
+	Command        string
+	Name           string
+	Description    string
+	Usage          string
+	Plugin         string
+	CommandHandler func(ctx context.Context, mmCommand *MMSlashCommand, args map[string]string) (*CommandResponse, error)
```
Comment (pfltdv) [inline model/command.go:11]: :heart_eyes: Sweet

**Review by pfltdv — APPROVED** [review]
:100:. Amazing work Julien. It is more structured and organised now. Thanks alot for doing this refactoring.

**Comment by JulienTant** [comment]
> 💯. Amazing work Julien. It is more structured and organised now. Thanks alot for doing this refactoring.

Thanks @pfltdv! I'm glad you like the changes!

**Pull request merged**